dev/AppNotifications/AppNotificationBuilder/AppNotificationBuilder.cpp

```diff
@@ -0,0 +1,372 @@
+            { AppNotificationSoundEvent::Alarm5, L"ms-winsoundevent:Notification.Looping.Alarm5"},
+            { AppNotificationSoundEvent::Alarm6, L"ms-winsoundevent:Notification.Looping.Alarm6"},
+            { AppNotificationSoundEvent::Alarm7, L"ms-winsoundevent:Notification.Looping.Alarm7"},
+            { AppNotificationSoundEvent::Alarm8, L"ms-winsoundevent:Notification.Looping.Alarm8"},
+            { AppNotificationSoundEvent::Alarm9, L"ms-winsoundevent:Notification.Looping.Alarm9"},
+            { AppNotificationSoundEvent::Alarm10, L"ms-winsoundevent:Notification.Looping.Alarm10"},
+            { AppNotificationSoundEvent::Call, L"ms-winsoundevent:Notification.Looping.Call"},
+            { AppNotificationSoundEvent::Call2, L"ms-winsoundevent:Notification.Looping.Call2"},
+            { AppNotificationSoundEvent::Call3, L"ms-winsoundevent:Notification.Looping.Call3"},
+            { AppNotificationSoundEvent::Call4, L"ms-winsoundevent:Notification.Looping.Call4"},
+            { AppNotificationSoundEvent::Call5, L"ms-winsoundevent:Notification.Looping.Call5"},
+            { AppNotificationSoundEvent::Call6, L"ms-winsoundevent:Notification.Looping.Call6"},
+            { AppNotificationSoundEvent::Call7, L"ms-winsoundevent:Notification.Looping.Call7"},
+            { AppNotificationSoundEvent::Call8, L"ms-winsoundevent:Notification.Looping.Call8"},
+            { AppNotificationSoundEvent::Call9, L"ms-winsoundevent:Notification.Looping.Call9"},
+            { AppNotificationSoundEvent::Call10, L"ms-winsoundevent:Notification.Looping.Call10"},
+        };
+
+        return c_soundEventMap[soundEvent];
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::AddArgument(hstring const& key, hstring const& value)
+    {
+        THROW_HR_IF(E_INVALIDARG, key.empty());
+
+        m_arguments.Insert(key, value);
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetTimeStamp(winrt::Windows::Foundation::DateTime const& value)
+    {
+        //winrt::hstring currentLanguage { ApplicationLanguages::Languages().First().Current() };
+        //// DateTimeFormatter dateTimeFormatter{ "longtime" };
+        //winrt::Windows::Foundation::DateTime::
+        //winrt::Windows::Foundation::DateTime { 2022, 1, 13, 16, 25, 30,  }
+        //m_timeStamp = timeStamp;
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::SetScenario(AppNotificationScenario const& value)
+    {
+        m_scenario = value;
+        return *this;
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::Builder::AppNotificationBuilder AppNotificationBuilder::AddText(hstring const& text)
+    {
+        THROW_HR_IF(E_INVALIDARG, m_textLines.size() == 3u);
+
+        m_textLines.push_back(L"<text>" + text + L"</text>");
```

Comment:
String concat is generally expensive. The cost likely adds up with the builder pattern. Can we use String format syntax instead?

Reply:
Yeah I’m looking into other options for string concat. Above I mentioned using fmt::format which builds a string with printf format. Is there an std function I can use for this?

Reply:
fmtlib is not std c++ right? if I understand correctly and it is simply an open sourced implementation built on top of printf. Other option is to use std::stringstream. 

Would you be able to measure performance for string concat Vs stringstream to see if the benefits warrant the change?

Reply:
[std::string.reserve()](https://cplusplus.com/reference/string/string/reserve/) is your friend. That preallocates a buffer size so if you know a typical worst case you can do this up front then concat e.g.

```
std::wstring s;
s.reserve(4096);
s.append(L"hello");
s.append(123);
s.append(foo);
```

Though `push_back()` may be more efficient than `append()`.

Reply:
I STRONGLY discourage bringing in fmtlib (or other libraries). Licensing, supply chain tracking, security review and other issues must be addressed for any dependency brought into Windows App SDK. Sufficient justification is needed to add any library.

We have access to Windows, the Standard C++ library, WIL, C++/WinRT and we'll have access to C++20 soon enough which brings `std::format()` and other goodies. I don't see the justification for fmtlib.

If you think there's sufficient justification feel free to make your case. We're currently working on refining the process for dependencies/tool version management so for now I'd suggest starting your proposal to @BenJKuhn, @MikeHillberg and myself and we ran rope in additional folks as appropriate.

Reply:
Fixed string concat issues